### ISSUE TITLE
sqlfluff fix incorrectly move `FROM` to the end of its previous line  

### ISSUE THREAD

**Issue by xg1990** [sqlfluff/sqlfluff#3399]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

For the below sql 
![image](https://user-images.githubusercontent.com/1443518/170423452-5a1e0c52-9412-48b3-a062-be4b3b63dc80.png)

If I ran `sqlfluff fix`

![image](https://user-images.githubusercontent.com/1443518/170423681-6700546c-a10e-4a53-86d3-4f995cf987df.png)

### Expected Behaviour

The keyword `FROM` should not be moved

### Observed Behaviour

Then it became 

![image](https://user-images.githubusercontent.com/1443518/170423550-33198c07-7acd-4dae-a336-cff43f9450c6.png)

Please note that the keyword `FROM` is moved to an incorrect place

### How to reproduce

```sql
SELECT
    a.*,
    b as col
FROM
    {{source("ids","shop")}} a;
```

just run sqlfluff fix with the above SQL

### Dialect

tsql

### Version

sqlfluff==0.13.2
sqlfluff-templater-dbt==0.9.4
dbt-core==1.0.0
dbt-extractor==0.4.0

### Configuration

```
[sqlfluff]
verbose = 0
nocolor = False
dialect = tsql
templater = jinja
# rules = L001,L002,L003,L004,L005,L006,L007,L008,L009,L010,L011,L012,L014,L029,L030,L039,L040,L050,L059,L062,L063,L065
exclude_rules = L031,L044
recurse = 0
output_line_length = 80
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
disable_noqa = False
# Comma separated list of file extensions to lint.
# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:indentation]
indented_joins = True
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = False

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
load_macros_from_path=./macros

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}.{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 80
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L007]  # Keywords
operator_new_lines = before

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

[sqlfluff:rules:L011]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L012]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L026]
force_enable = False

[sqlfluff:rules:L028]
force_enable = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = upper

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = upper

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L052]  # Semi-colon formatting approach.
multiline_newline = False
require_final_semicolon = False

[sqlfluff:rules:L054]  # GROUP BY/ORDER BY column references.
group_by_and_order_by_style = consistent

[sqlfluff:rules:L057]  # Special characters in identifiers
unquoted_identifiers_policy = all
quoted_identifiers_policy = all
allow_space_in_identifier = False
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by xg1990** [comment]
No such problem with `sqlfluff==0.13.1`, this bug is probably introduced in `0.13.2`

**Comment by barrywhart** [comment]
This is an odd bug, because the corruption doesn't appear to be caused by any particular rule. I tweaked the test SQL a few different ways so different (individual) rules fired, and the problem always occurred, e.g.:
```
select
    a.*,
    a.b AS col
FROM
    {{source("ids","shop")}} AS a;
```

(In the case above, the only fix is to capitalize `select`. Note this is a very different area of the code than the `FROM`.

**Comment by barrywhart** [comment]
Things I've discovered:
* The issue does not occur in `ansi` dialect.
* My tests showed a different result than the original report: I see the issue in 0.13.2 and 0.13.1, but not in 0.13.0.

@tunetheweb: Are you aware of any `tsql` dialect changes in the couple of releases that might've caused this? I don't know the root cause, but I'm wondering if the code is being parsed slightly differently, e.g. whitespace is being positioned differently.

**Comment by barrywhart** [comment]
I used `git bisect` to narrow it down to this PR: #3283: Enhance apply_fixes() to automatically fix violations of can_start_end_non_code. This was my PR, so I'll look into fixing it.
